### DISCUSSION TITLE
Injection/production in one grid element

### DISCUSSION THREAD

**Discussion by pierre-jeanne** [idaholab/moose#22772] [Q&A General] [answered]
Hello Moose community,

I try to simulate water injection and production into two grid elements, with PorousFlowSink:
```
 [injection]
    type = PorousFlowSink
    boundary = '8'
    variable = porepressure
    flux_function = -1E-2
 [] 
'
and 
'
 [production]
    type = PorousFlowSink
    boundary = '9'
    variable = porepressure
    flux_function = 1E-2
  []
[]
```
8 and 9 are made of one single element respectively. Is it possible to do that?
 Here is the message error :
```
*** ERROR ***
path 89:(BCs/injection/boundary):  
the following side sets (ids) do not exist on the mesh: 8
```
Ideally, I would like to impose a fixed pore pressure in one grid element and impose a fixed production rate (kg/s) in another grid element.
Do you have any advice on how to do that?
Thanks!!!

**Comment by GiudGiud** [comment]
Hello

When you open the mesh do you see boundary 8?
A single side of an element should be fine

Guillaume

**Reply by pierre-jeanne** [reply]
yes, I can see it in ParaView.However, when I tried another subdomain_id, (e.g. boundary = '2'), the simulation started without an error message...

**Reply by GiudGiud** [reply]
it should be a boundary id not a subdomain id.

In paraview do you see a sideset or a nodeset?

**Reply by pierre-jeanne** [reply]
yes in paraview I can see them.
I have re-generated the mesh using 'show_info = true'. I can see that the subdomain exists with 8 active nodes.
Where can I find the list of the boundary id? is there a command line to get the characteristic of the mesh (nodes coordinates, center of the block, boundary coordinate etc...)? i looked for that but I found nothing.

**Reply by GiudGiud** [reply]
You keep referring to 8 as a subdomain, but it needs to be a sideset not a subdomain. 

show_info=true shows all the sidesets in the mesh. 
Could you please paste that here?

**Reply by pierre-jeanne** [reply]
"you keep referring to 8 as a subdomain, but it needs to be a sideset not a subdomain."
yes, I understand.

Maybe I had it wrong, so let me reformulate my question.
I have a 2d horizontal model. I want to inject (or produce) in the middle of the MESH. 
I used TOUGH2 a lot, and in TOUGH2 you change the Pressure within one grid element.
In MOOSE, I understand you have to fix the pressure at the boundary between two grid elements, is that correct?
In this case, how can you find the boundaries id?  

Here is the results get with show_info

CartesianMeshGenerator 'cmg':  Mesh Information:
CartesianMeshGenerator 'cmg':   elem_dimensions()={3}
CartesianMeshGenerator 'cmg':   spatial_dimension()=3
CartesianMeshGenerator 'cmg':   n_nodes()=7656
CartesianMeshGenerator 'cmg':     n_local_nodes()=7656
CartesianMeshGenerator 'cmg':   n_elem()=3705
CartesianMeshGenerator 'cmg':     n_local_elem()=3705
CartesianMeshGenerator 'cmg':     n_active_elem()=3705
CartesianMeshGenerator 'cmg':   n_subdomains()=9
CartesianMeshGenerator 'cmg':   n_elemsets()=0
CartesianMeshGenerator 'cmg':   n_partitions()=1
CartesianMeshGenerator 'cmg':   n_processors()=1
CartesianMeshGenerator 'cmg':   n_threads()=1
CartesianMeshGenerator 'cmg':   processor_id()=0
CartesianMeshGenerator 'cmg':   is_prepared()=true
CartesianMeshGenerator 'cmg':   is_replicated()=true
CartesianMeshGenerator 'cmg':
CartesianMeshGenerator 'cmg':  Mesh Bounding Box:
CartesianMeshGenerator 'cmg':   Minimum: (x,y,z)=(       0,        0,        0)
CartesianMeshGenerator 'cmg':   Maximum: (x,y,z)=( 2449.97,  3063.61,       30)
CartesianMeshGenerator 'cmg':   Delta:   (x,y,z)=( 2449.97,  3063.61,       30)
CartesianMeshGenerator 'cmg':
CartesianMeshGenerator 'cmg':  Mesh Element Type(s):
CartesianMeshGenerator 'cmg':   HEX8
CartesianMeshGenerator 'cmg':
CartesianMeshGenerator 'cmg':  Mesh Nodesets:
CartesianMeshGenerator 'cmg':   Nodeset 0 (back), 3828 nodes
CartesianMeshGenerator 'cmg':    Bounding box minimum: (x,y,z)=(       0,        0,        0)
CartesianMeshGenerator 'cmg':    Bounding box maximum: (x,y,z)=( 2449.97,  3063.61,        0)
CartesianMeshGenerator 'cmg':    Bounding box delta: (x,y,z)=( 2449.97,  3063.61,        0)
CartesianMeshGenerator 'cmg':   Nodeset 1 (bottom), 132 nodes
CartesianMeshGenerator 'cmg':    Bounding box minimum: (x,y,z)=(       0,        0,        0)
CartesianMeshGenerator 'cmg':    Bounding box maximum: (x,y,z)=( 2449.97,        0,       30)
CartesianMeshGenerator 'cmg':    Bounding box delta: (x,y,z)=( 2449.97,        0,       30)
CartesianMeshGenerator 'cmg':   Nodeset 2 (right), 116 nodes
CartesianMeshGenerator 'cmg':    Bounding box minimum: (x,y,z)=( 2449.97,        0,        0)
CartesianMeshGenerator 'cmg':    Bounding box maximum: (x,y,z)=( 2449.97,  3063.61,       30)
CartesianMeshGenerator 'cmg':    Bounding box delta: (x,y,z)=(       0,  3063.61,       30)
CartesianMeshGenerator 'cmg':   Nodeset 3 (top), 132 nodes
CartesianMeshGenerator 'cmg':    Bounding box minimum: (x,y,z)=(       0,  3063.61,        0)
CartesianMeshGenerator 'cmg':    Bounding box maximum: (x,y,z)=( 2449.97,  3063.61,       30)
CartesianMeshGenerator 'cmg':    Bounding box delta: (x,y,z)=( 2449.97,        0,       30)
CartesianMeshGenerator 'cmg':   Nodeset 4 (left), 116 nodes
CartesianMeshGenerator 'cmg':    Bounding box minimum: (x,y,z)=(       0,        0,        0)
CartesianMeshGenerator 'cmg':    Bounding box maximum: (x,y,z)=(       0,  3063.61,       30)
CartesianMeshGenerator 'cmg':    Bounding box delta: (x,y,z)=(       0,  3063.61,       30)
CartesianMeshGenerator 'cmg':   Nodeset 5 (front), 3828 nodes
CartesianMeshGenerator 'cmg':    Bounding box minimum: (x,y,z)=(       0,        0,       30)
CartesianMeshGenerator 'cmg':    Bounding box maximum: (x,y,z)=( 2449.97,  3063.61,       30)
CartesianMeshGenerator 'cmg':    Bounding box delta: (x,y,z)=( 2449.97,  3063.61,        0)
CartesianMeshGenerator 'cmg':
CartesianMeshGenerator 'cmg':  Mesh Sidesets:
CartesianMeshGenerator 'cmg':   Sideset 0 (back), 3705 sides (QUAD4), 3705 elems (HEX8), 3828 nodes
CartesianMeshGenerator 'cmg':    Side volume: 7.50575e+06
CartesianMeshGenerator 'cmg':    Bounding box minimum: (x,y,z)=(       0,        0,        0)
CartesianMeshGenerator 'cmg':    Bounding box maximum: (x,y,z)=( 2449.97,  3063.61,        0)
CartesianMeshGenerator 'cmg':    Bounding box delta: (x,y,z)=( 2449.97,  3063.61,        0)
CartesianMeshGenerator 'cmg':   Sideset 1 (bottom), 65 sides (QUAD4), 65 elems (HEX8), 132 nodes
CartesianMeshGenerator 'cmg':    Side volume: 73499.1
CartesianMeshGenerator 'cmg':    Bounding box minimum: (x,y,z)=(       0,        0,        0)
CartesianMeshGenerator 'cmg':    Bounding box maximum: (x,y,z)=( 2449.97,        0,       30)
CartesianMeshGenerator 'cmg':    Bounding box delta: (x,y,z)=( 2449.97,        0,       30)
CartesianMeshGenerator 'cmg':   Sideset 2 (right), 57 sides (QUAD4), 57 elems (HEX8), 116 nodes
CartesianMeshGenerator 'cmg':    Side volume: 91908.3
CartesianMeshGenerator 'cmg':    Bounding box minimum: (x,y,z)=( 2449.97,        0,        0)
CartesianMeshGenerator 'cmg':    Bounding box maximum: (x,y,z)=( 2449.97,  3063.61,       30)
CartesianMeshGenerator 'cmg':    Bounding box delta: (x,y,z)=(       0,  3063.61,       30)
CartesianMeshGenerator 'cmg':   Sideset 3 (top), 65 sides (QUAD4), 65 elems (HEX8), 132 nodes
CartesianMeshGenerator 'cmg':    Side volume: 73499.1
CartesianMeshGenerator 'cmg':    Bounding box minimum: (x,y,z)=(       0,  3063.61,        0)
CartesianMeshGenerator 'cmg':    Bounding box maximum: (x,y,z)=( 2449.97,  3063.61,       30)
CartesianMeshGenerator 'cmg':    Bounding box delta: (x,y,z)=( 2449.97,        0,       30)
CartesianMeshGenerator 'cmg':   Sideset 4 (left), 57 sides (QUAD4), 57 elems (HEX8), 116 nodes
CartesianMeshGenerator 'cmg':    Side volume: 91908.3
CartesianMeshGenerator 'cmg':    Bounding box minimum: (x,y,z)=(       0,        0,        0)
CartesianMeshGenerator 'cmg':    Bounding box maximum: (x,y,z)=(       0,  3063.61,       30)
CartesianMeshGenerator 'cmg':    Bounding box delta: (x,y,z)=(       0,  3063.61,       30)
CartesianMeshGenerator 'cmg':   Sideset 5 (front), 3705 sides (QUAD4), 3705 elems (HEX8), 3828 nodes
CartesianMeshGenerator 'cmg':    Side volume: 7.50575e+06
CartesianMeshGenerator 'cmg':    Bounding box minimum: (x,y,z)=(       0,        0,       30)
CartesianMeshGenerator 'cmg':    Bounding box maximum: (x,y,z)=( 2449.97,  3063.61,       30)
CartesianMeshGenerator 'cmg':    Bounding box delta: (x,y,z)=( 2449.97,  3063.61,        0)
CartesianMeshGenerator 'cmg':
CartesianMeshGenerator 'cmg':  Mesh Edgesets:
CartesianMeshGenerator 'cmg':   None
CartesianMeshGenerator 'cmg':
CartesianMeshGenerator 'cmg':  Mesh Subdomains:
CartesianMeshGenerator 'cmg':   Subdomain 1: 3078 elems (HEX8, 3078 active), 6496 active nodes
CartesianMeshGenerator 'cmg':    Volume: 2.25171e+08
CartesianMeshGenerator 'cmg':    Bounding box minimum: (x,y,z)=(       0,        0,        0)
CartesianMeshGenerator 'cmg':    Bounding box maximum: (x,y,z)=( 2449.97,  3063.61,       30)
CartesianMeshGenerator 'cmg':    Bounding box delta: (x,y,z)=( 2449.97,  3063.61,       30)
CartesianMeshGenerator 'cmg':   Subdomain 2: 55 elems (HEX8, 55 active), 232 active nodes
CartesianMeshGenerator 'cmg':    Volume: 91.8963
CartesianMeshGenerator 'cmg':    Bounding box minimum: (x,y,z)=( 1224.98,        0,        0)
CartesianMeshGenerator 'cmg':    Bounding box maximum: (x,y,z)=( 1224.98,  3063.61,       30)
CartesianMeshGenerator 'cmg':    Bounding box delta: (x,y,z)=(   0.001,  3063.61,       30)
CartesianMeshGenerator 'cmg':   Subdomain 3: 114 elems (HEX8, 114 active), 464 active nodes
CartesianMeshGenerator 'cmg':    Volume: 183.817
CartesianMeshGenerator 'cmg':    Bounding box minimum: (x,y,z)=( 1224.98,        0,        0)
CartesianMeshGenerator 'cmg':    Bounding box maximum: (x,y,z)=( 1224.99,  3063.61,       30)
CartesianMeshGenerator 'cmg':    Bounding box delta: (x,y,z)=(   0.003,  3063.61,       30)
CartesianMeshGenerator 'cmg':   Subdomain 4: 114 elems (HEX8, 114 active), 464 active nodes
CartesianMeshGenerator 'cmg':    Volume: 183.817
CartesianMeshGenerator 'cmg':    Bounding box minimum: (x,y,z)=( 1224.98,        0,        0)
CartesianMeshGenerator 'cmg':    Bounding box maximum: (x,y,z)=( 1224.99,  3063.61,       30)
CartesianMeshGenerator 'cmg':    Bounding box delta: (x,y,z)=(   0.005,  3063.61,       30)
CartesianMeshGenerator 'cmg':   Subdomain 5: 114 elems (HEX8, 114 active), 464 active nodes
CartesianMeshGenerator 'cmg':    Volume: 183.817
CartesianMeshGenerator 'cmg':    Bounding box minimum: (x,y,z)=( 1224.98,        0,        0)
CartesianMeshGenerator 'cmg':    Bounding box maximum: (x,y,z)=( 1224.99,  3063.61,       30)
CartesianMeshGenerator 'cmg':    Bounding box delta: (x,y,z)=(   0.007,  3063.61,       30)
CartesianMeshGenerator 'cmg':   Subdomain 6: 114 elems (HEX8, 114 active), 464 active nodes
CartesianMeshGenerator 'cmg':    Volume: 183.817
CartesianMeshGenerator 'cmg':    Bounding box minimum: (x,y,z)=( 1224.98,        0,        0)
CartesianMeshGenerator 'cmg':    Bounding box maximum: (x,y,z)=( 1224.99,  3063.61,       30)
CartesianMeshGenerator 'cmg':    Bounding box delta: (x,y,z)=(   0.009,  3063.61,       30)
CartesianMeshGenerator 'cmg':   Subdomain 7: 114 elems (HEX8, 114 active), 464 active nodes
CartesianMeshGenerator 'cmg':    Volume: 183.817
CartesianMeshGenerator 'cmg':    Bounding box minimum: (x,y,z)=( 1224.98,        0,        0)
CartesianMeshGenerator 'cmg':    Bounding box maximum: (x,y,z)=( 1224.99,  3063.61,       30)
CartesianMeshGenerator 'cmg':    Bounding box delta: (x,y,z)=(   0.011,  3063.61,       30)
CartesianMeshGenerator 'cmg':   Subdomain 8: 1 elems (HEX8, 1 active), 8 active nodes
CartesianMeshGenerator 'cmg':    Volume: 0.006
CartesianMeshGenerator 'cmg':    Bounding box minimum: (x,y,z)=( 1224.98,  1295.81,        0)
CartesianMeshGenerator 'cmg':    Bounding box maximum: (x,y,z)=( 1224.98,  1296.01,       30)
CartesianMeshGenerator 'cmg':    Bounding box delta: (x,y,z)=(   0.001,      0.2,       30)
CartesianMeshGenerator 'cmg':   Subdomain 9: 1 elems (HEX8, 1 active), 8 active nodes
CartesianMeshGenerator 'cmg':    Volume: 0.006
CartesianMeshGenerator 'cmg':    Bounding box minimum: (x,y,z)=( 1224.98,  1767.61,        0)
CartesianMeshGenerator 'cmg':    Bounding box maximum: (x,y,z)=( 1224.98,  1767.81,       30)
CartesianMeshGenerator 'cmg':    Bounding box delta: (x,y,z)=(   0.001,      0.2,       30)
CartesianMeshGenerator 'cmg':   Global mesh volume = 2.25172e+08

**Reply by pierre-jeanne** [reply]
I think I know, I should use 'ParsedGenerateSideset' when generating the mesh to select and name the side where I want to apply my BC.
Is that correct?

**Reply by GiudGiud** [reply]
I ll let @WilkAndy comment on how to inject pressure. 

With regards to sidesets, under
```
CartesianMeshGenerator 'cmg': Mesh Sidesets:
```
you can see that there are only 6 sidesets, and none of them have an ID of 8.

You may use meshgenerators, such as the SidesetBetweenSubdomainGenerator to add new sidesets
https://mooseframework.inl.gov/source/meshgenerators/SideSetsBetweenSubdomainsGenerator.html

**Reply by GiudGiud** [reply]
ParsedGenerateSideset would work too

**Comment by WilkAndy** [comment, marked as answer]
@pierre-jeanne - i think you might be using the wrong PorousFlow objects.  The `BCs` objects like `PorousFlowSink` are usually used on boundaries of your mesh, eg, on the entire left-hand edge of the mesh, in your case.   Instead, I believe what you want is two `DiracKernels`, which inject and produce at a single point (into a single element, for instance).  Have a look at https://mooseframework.inl.gov/modules/porous_flow/sinks.html , which details quite a variety of point (and line) sinks/sources available.  You probably want https://mooseframework.inl.gov/source/dirackernels/PorousFlowSquarePulsePointSource.html or https://mooseframework.inl.gov/source/dirackernels/PorousFlowPointSourceFromPostprocessor.html .

**Reply by pierre-jeanne** [reply]
Hello Andy,

thank you for your answer. You are right, the two `Dirackernels`  can do what I was looking for.
I used `PorousFlowSquarePulsePointSource` to simulate injection and production. My simulation run, there is no error message but the pressure doesn't change. I am doing something wrong but I don't know what. Do you mind having a look at the code below?

Thanks.


[GlobalParams]
  PorousFlowDictator = dictator
[]

############################################################
[Variables]
  [porepressure]
    initial_condition = 20.0E6
  []
[]
#####################################################
[PorousFlowBasicTHM]
  porepressure = porepressure    # There is just one variable — the porepressure
  coupling_type = Hydro          # there is no coupling with heat or mechanics.
  gravity = '0 0 -9.81'          # gravity is vertical
  fp = the_simple_fluid          # Most PorousFlow simulations require fluid properties to be supplied.
[]
############################################################
# The SimpleFluidProperties are used, which assume a constant fluid bulk modulus and viscosity:
[FluidProperties]
  [the_simple_fluid]
    type = SimpleFluidProperties
    bulk_modulus = 2E9
    viscosity = 1.0E-3
    density0 = 1000.0
  []
[]
#####################################################
[Kernels]
  [mass0]
    type = PorousFlowMassTimeDerivative
    variable = porepressure
  []
[]
############################################################
[BCs]
  [./left]                              # arbitrary name
    type = DirichletBC
    variable = porepressure
    value = 20E6
    boundary = left  # This must match a named boundary in the mesh file
  [../]

  [./right]                             # arbitrary name
    type = DirichletBC
    variable = porepressure
    value = 20E6
    boundary = right # This must match a named boundary in the mesh file
  [../]

  [./front]                              # arbitrary name
    type = DirichletBC
    variable = porepressure
    value = 20E6
    boundary = front # This must match a named boundary in the mesh file
  [../]

  [./back]                              # arbitrary name
    type = DirichletBC
    variable = porepressure
    value = 20E6
    boundary = back # This must match a named boundary in the mesh file
  [../]
[]

############################################################
# The porosity, Biot modulus and permeability are defined through the Materials block:
[Materials]
  [porosity]
    type = PorousFlowPorosityConst
    block = '1 3 4 5 6 7'
    porosity = 0.01           
  []
  [porosity_fracture]
    type = PorousFlowPorosityConst
    block = '2 8 9'
    porosity = 0.05           
  []
  [permeability_intact_rock]
    type = PorousFlowPermeabilityConst
    block = '1 3 4 5 6 7'
    permeability = '1E-16 0 0   0 1E-16 0   0 0 1E-16'
  []
  [permeability_fracture]
    type = PorousFlowPermeabilityConst
    block = '2 8 9'
    permeability = '1E-11 0 0   0 1E-11 0   0 0 1E-11'
  []
  [biot_modulus]
    type = PorousFlowConstantBiotModulus
    block = '1 2 3 4 5 6 7 8 9'
    fluid_bulk_modulus = 2E9
    biot_coefficient = 1.0
  []
[]
#####################################################
[DiracKernels]
  [sink1]
    type = PorousFlowSquarePulsePointSource
    block = '8'
    point = '1224.985 1296.005 0'
    start_time = 604800
    mass_flux = -0.353
    variable = porepressure
    point_not_found_behavior = ERROR
  []

  [source]
    type = PorousFlowSquarePulsePointSource
    block = '9'
    point = '1224.985 1767.805 0'
    mass_flux = 9.353
    variable = porepressure
    point_not_found_behavior = ERROR
  []
[]
############################################################
[Postprocessors]
  [./p0]
    type = PointValue
    outputs = csv
    point = '1224.985 1296.005 0'
    variable = porepressure
  [../]
  [./p1]
    type = PointValue
    outputs = csv
    point = '1224.985 1767.805 0'
    variable = porepressure
  [../]
[]
############################################################
[Executioner]
  type = Transient
  solve_type = Newton
  start_time = 0.0
  end_time = 7.884e8
  [TimeStepper]
    type = IterationAdaptiveDT
    optimal_iterations = 1e7
    dt = 1E2
  []
[]

############################################################
[Outputs]
  execute_on = 'timestep_end'
  exodus = true
  [./csv]
    type = CSV
  [../]
[]

**Reply by WilkAndy** [reply]
You don't need the `Kernels` block, since `PorousFlowBasicTHM` automatically adds it.   Just be careful with the `multiply_by_density` flag: see https://mooseframework.inl.gov/source/actions/PorousFlowBasicTHM.html for a discussion.  Eg, you when `multiply_by_density = false` (which is the default) you will actually be injecting in m^3/s, not in kg/s.

I'm about to board a plane, sorry for the rushed response.   What do you mean by "the pressure doesn't change" - what do the nonlinear iterations look like?  If they're all "residual = 0" then the porepressure won't have changed, but if you get >=1 nonlinear iteration per timestep then the pressure must have changed, but your postprocessing (visualisation) must be wrong.

**Reply by pierre-jeanne** [reply]
Thank you for your reply! yes the residuals are equal to 0.

**Reply by WilkAndy** [reply]
Hmmm, i can't see anything obviously wrong - i must be missing something.  i can't run your input file because i don't have the mesh.  Can you please:
- use `multiply_by_density = true`
- remove the `Kernels` block
- make a boring `GeneratedMesh` and use that instead, just in case it's a mesh problem.   You don't need multiple blocks - just one block that covers your domain is sufficient to see if "residual = 0" or not.

a

**Reply by pierre-jeanne** [reply]
Ok it was a problem with the mesh. I appreciate your help!

I had a 3D mesh but with only one layer in the z-direction (and it didn't work), I added a layer above and below (3 layers in total),  and it worked.
Do you know why? Is it because the nodes where I simulated injection and production were on the upper and lower boundaries (now they are in the middle of the mesh)? 

Pierre

**Reply by WilkAndy** [reply]
@pierre-jeanne , it's good to know that it works now, but i think we need to explore further, because the situation you describe (with 1 mesh layer) should have worked fine.   Could you please try using your original 1-layer mesh, and move your `point`s just a little bit so they're definitely inside the mesh domain?

**Reply by pierre-jeanne** [reply]
Hello Andy,
I  changed the coordinate to locate the injection and production points in the middle of the grid element, but it didn't fix the problem.

You can try it by yourself.  I copy and past the two codes:
- the first one (1 layer) ran without an error message but injection and production do not 'work'.
- the 2nd (3 layers) ran fine. 

First one: 
[Mesh]
  [cmg]
    type = CartesianMeshGenerator
    dim = 3
    dx = '
    1 1 1 1 1
    '
    dy = '
    1 1 1 1 1
    '
    dz = '1'
    subdomain_id = '
    1 1 1 1 1
    1 1 8 1 1
    1 1 1 1 1
    1 1 9 1 1
    1 1 1 1 1
    '
    show_info = True
  []
[]

[GlobalParams]
  PorousFlowDictator = dictator
[]
[Variables]
  [porepressure]
    initial_condition = 20.0E6
  []
[]
[PorousFlowBasicTHM]
  porepressure = porepressure    # There is just one variable — the porepressure
  coupling_type = Hydro          # there is no coupling with heat or mechanics.
  gravity = '0 0 -9.81'          # gravity is vertical
  multiply_by_density = true     # to inject kg/s 
  fp = the_simple_fluid          # Most PorousFlow simulations require fluid properties to be supplied.
[]
[FluidProperties]
  [the_simple_fluid]
    type = SimpleFluidProperties
    bulk_modulus = 2E9
    viscosity = 1.0E-3
    density0 = 1000.0
  []
[]
[BCs]
  [./left] # arbitrary user-chosen name
    type = DirichletBC
    variable = porepressure
    value = 20E6
    boundary = left # This must match a named boundary in the mesh file
  [../]

  [./right] # arbitrary user-chosen name
    type = DirichletBC
    variable = porepressure
    value = 20E6
    boundary = right # This must match a named boundary in the mesh file
  [../]

  [./front] # arbitrary user-chosen name
    type = DirichletBC
    variable = porepressure
    value = 20E6
    boundary = front # This must match a named boundary in the mesh file
  [../]

  [./back] # arbitrary user-chosen name
    type = DirichletBC
    variable = porepressure
    value = 20E6
    boundary = back # This must match a named boundary in the mesh file
  [../]
[]
[Materials]
  [porosity]
    type = PorousFlowPorosityConst
    block = '1 8 9'
    porosity = 0.01           
  []
  [permeability]
    type = PorousFlowPermeabilityConst
    block = '1 8 9'
    permeability = '1E-16 0 0   0 1E-16 0   0 0 1E-16'
  []
  [biot_modulus]
    type = PorousFlowConstantBiotModulus
    block = '1 8 9'
    fluid_bulk_modulus = 2E9
    biot_coefficient = 1.0
  []
[]
[DiracKernels]
  [sink1]
    type = PorousFlowSquarePulsePointSource
    block = '8'
    point = '2 1 0'
    start_time = 604800
    mass_flux = -0.379
    variable = porepressure
    point_not_found_behavior = ERROR
  []

  [source]
    type = PorousFlowSquarePulsePointSource
    block = '9'
    point = '2.5 3.5 0.5'
    mass_flux = 0.379
    variable = porepressure
    point_not_found_behavior = ERROR
  []
[]
[Executioner]
  type = Transient
  solve_type = Newton
  start_time = 0.0
  end_time = 7.884e8
  nl_abs_tol = 1e-6

  [TimeStepper]
    type = IterationAdaptiveDT
    optimal_iterations = 1e7
    dt = 1E2
  []
[]
[Outputs]
  execute_on = 'timestep_end'
  exodus = true
  [./csv]
    type = CSV
  [../]
[]
#########################
the second code:
[Mesh]
  [cmg]
    type = CartesianMeshGenerator
    dim = 3
    dx = '
    1 1 1 1 1
    '
    dy = '
    1 1 1 1 1
    '
    dz = '1 1 1'
    subdomain_id = '
    1 1 1 1 1
    1 1 1 1 1
    1 1 1 1 1
    1 1 1 1 1
    1 1 1 1 1

    1 1 1 1 1
    1 1 8 1 1
    1 1 1 1 1
    1 1 9 1 1
    1 1 1 1 1

    1 1 1 1 1
    1 1 1 1 1
    1 1 1 1 1
    1 1 1 1 1
    1 1 1 1 1
    '
    show_info = True
  []
[]
[GlobalParams]
  PorousFlowDictator = dictator
[]
[Variables]
  [porepressure]
    initial_condition = 20.0E6
  []
[]
[PorousFlowBasicTHM]
  porepressure = porepressure    # There is just one variable — the porepressure
  coupling_type = Hydro          # there is no coupling with heat or mechanics.
  gravity = '0 0 -9.81'          # gravity is vertical
  multiply_by_density = true     # to inject kg/s 
  fp = the_simple_fluid          # Most PorousFlow simulations require fluid properties to be supplied.
[]
[FluidProperties]
  [the_simple_fluid]
    type = SimpleFluidProperties
    bulk_modulus = 2E9
    viscosity = 1.0E-3
    density0 = 1000.0
  []
[]
[BCs]
  [./left] # arbitrary user-chosen name
    type = DirichletBC
    variable = porepressure
    value = 20E6
    boundary = left # This must match a named boundary in the mesh file
  [../]

  [./right] # arbitrary user-chosen name
    type = DirichletBC
    variable = porepressure
    value = 20E6
    boundary = right # This must match a named boundary in the mesh file
  [../]

  [./front] # arbitrary user-chosen name
    type = DirichletBC
    variable = porepressure
    value = 20E6
    boundary = front # This must match a named boundary in the mesh file
  [../]

  [./back] # arbitrary user-chosen name
    type = DirichletBC
    variable = porepressure
    value = 20E6
    boundary = back # This must match a named boundary in the mesh file
  [../]
[]
[Materials]
  [porosity]
    type = PorousFlowPorosityConst
    block = '1 8 9'
    porosity = 0.01           
  []
  [permeability]
    type = PorousFlowPermeabilityConst
    block = '1 8 9'
    permeability = '1E-16 0 0   0 1E-16 0   0 0 1E-16'
  []
  [biot_modulus]
    type = PorousFlowConstantBiotModulus
    block = '1 8 9'
    fluid_bulk_modulus = 2E9
    biot_coefficient = 1.0
  []
[]
[DiracKernels]
  [sink1]
    type = PorousFlowSquarePulsePointSource
    block = '8'
    point = '2.5 1.5 1.5'
    start_time = 604800
    mass_flux = -0.379
    variable = porepressure
    point_not_found_behavior = ERROR
  []

  [source]
    type = PorousFlowSquarePulsePointSource
    block = '9'
    point = '2.5 3.5 1.5'
    mass_flux = 0.379
    variable = porepressure
    point_not_found_behavior = ERROR
  []
[]
[Executioner]
  type = Transient
  solve_type = Newton
  start_time = 0.0
  end_time = 7.884e8
  nl_abs_tol = 1e-6

  [TimeStepper]
    type = IterationAdaptiveDT
    optimal_iterations = 1e7
    dt = 1E2
  []
[]
[Outputs]
  execute_on = 'timestep_end'
  exodus = true
[]


Best,

Pierre

**Reply by WilkAndy** [reply]
Great, thanks for this, @pierre-jeanne - i hope to try this in the next 12 hours.

**Reply by WilkAndy** [reply]
OK, the solution turned out to be super easy.   Your `BCs` are wrong: they are fixing the porepressure on **zmax** and **zmin** (this is the `boundary = front` and `boundary = back` `BCs`).   So, there are no degrees of freedom for MOOSE to solve for.  Physically: you are injecting water, but immediately removing it from the zmax and zmin boundaries.  Ooops!

By the way, you can simulate this example using a 2D mesh - more computationally efficient.

**Reply by pierre-jeanne** [reply]
Yes it is working thanks. 

I am a little bit confused it is a 3d model with 6 boundaries: top, bottom, left, right, front, and back.  but when I fix the pore pressure on the front and back boundaries it also fixed the pore pressure on the top and bottom boundaries, why?

it is true that I didn't define the bottom and top BCs, because I made the assumption that when no the BCs are defined by default the boundaries are impermeable (it is like this in TOUGH2). Is it wrong? and is it why the front and back BCs are applied to the top and bottom BCs?

I simulated a 1 layer mesh instead of a 2d mesh because I wanted to compare my simulation results with the result obtained in TOUGH2 (with the exact same mesh).

Thanks again for your help and for taking the time to answer my question.
Best,
Piere

**Reply by WilkAndy** [reply]
Your confusion comes from the naming convention only.  You think that the word "top" means "boundary with the maximum z values".  However, the person that made the mesh (or created the MOOSE mesh generator to make that mesh) thinks that the word "front" means "boundary with the maximum z values".   In my mind, the naming convention is confusing, but we just have to accept that that's how it is.

To verify this, load the mesh into paraview.  When you load the mesh, you can choose to load "blocks" (3D elements in your case) and "sidesets" (front, back, top, etc).  Load the "front" sideset only, and you'll see that it is the boundary with maximum z values.

(I believe that TOUGH2 is a finite-volume solver, in which case a MOOSE 2D mesh will actually be more like the TOUGH2 simulation, but that's probably not really important to you.)

a
